File: testsuite/Jogl4Examples/ShadersExample/src/model/ShadersModel.java
package model;

import java.io.File;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.util.ArrayList;
import java.util.List;

import vsdk.toolkit.common.color.ColorRgb;
import vsdk.toolkit.environment.material.RendererConfiguration;
import vsdk.toolkit.common.linealAlgebra.Matrix4x4;
import vsdk.toolkit.common.linealAlgebra.Vector3D;
import vsdk.toolkit.environment.camera.Camera;
import vsdk.toolkit.environment.light.Light;
import vsdk.toolkit.environment.light.LightType;
import vsdk.toolkit.environment.material.SimpleMaterial;
import vsdk.toolkit.environment.material.MicroFacetedMaterial;
import vsdk.toolkit.environment.geometry.volume.Sphere;
import vsdk.toolkit.gui.CameraController;
import vsdk.toolkit.gui.CameraControllerOrbiter;
import vsdk.toolkit.gui.RendererConfigurationController;
import vsdk.toolkit.io.image.ImagePersistence;
import vsdk.toolkit.media.IndexedColorImageUncompressed;
import vsdk.toolkit.media.NormalMap;
import vsdk.toolkit.media.RGBImageUncompressed;

public class ShadersModel
{
    private static final int MIN_SPHERE_MERIDIANS = 12;
    private static final int MIN_SPHERE_PARALLELS = 8;
    private static final Vector3D DEFAULT_BUMP_SCALE = new Vector3D(1.0, 1.0, 1.0);
    private static final String MICROFACET_CSV_FILE =
        "../../../etc/materials/microFacetMAterials.csv";
    private static final String COOK_TORRANCE_MATERIAL_NAME = "Copper";

    private Camera camera;
    private CameraController cameraController;
    private RendererConfiguration quality;
    private RendererConfigurationController qualityController;
    private Sphere sphere;
    private Light light;
    private SimpleMaterial material;
    private MicroFacetedMaterial cookTorranceCopperMaterial;
    private List<String> cookTorranceMaterialNames;
    private int cookTorranceMaterialIndex;
    private RGBImageUncompressed textureMap;
    private RGBImageUncompressed bumpMapHeightRgb;
    private RGBImageUncompressed softwareFrameImage;
    private ShaderOperationMode renderingMode;
    private boolean showHud;
    private boolean animationEnabled;
    private boolean lightAnimationEnabled;
    private double sphereRotationAngleRadians;
    private int sphereMeridians;
    private int sphereParallels;

    public static ShadersModel createDefault()
    {
        ShadersModel model = new ShadersModel();
        model.initializeDefaults();
        return model;
    }

    private void initializeDefaults()
    {
        camera = new Camera();
        camera.setPosition(new Vector3D(0, -4, 0));
        Matrix4x4 rotation = new Matrix4x4().eulerAnglesRotation(Math.toRadians(90.0), 0, 0);
        camera.setRotation(rotation);
        camera.setFov(30.0);

        cameraController = new CameraControllerOrbiter(camera);

        quality = new RendererConfiguration();
        quality.setTexture(true);
        quality.setBumpMap(true);
        quality.setShadingType(RendererConfiguration.SHADING_TYPE_PHONG);
        qualityController = new RendererConfigurationController(quality);

        sphere = new Sphere(1.0);

        light = new Light(LightType.POINT, new Vector3D(1, -3, 1), new ColorRgb(1, 1, 1));
        light.setId(0);

        material = new SimpleMaterial();
        material = material.withAmbient(new ColorRgb(0.1, 0.1, 0.1));
        material = material.withDiffuse(new ColorRgb(1, 1, 1));
        material = material.withSpecular(new ColorRgb(1, 1, 1));
        material = material.withPhongExponent(40);
        cookTorranceCopperMaterial = new MicroFacetedMaterial(
            MICROFACET_CSV_FILE,
            COOK_TORRANCE_MATERIAL_NAME);
        cookTorranceMaterialNames = loadMicroFacetMaterialNames(MICROFACET_CSV_FILE);
        cookTorranceMaterialIndex = indexOfMaterialName(
            cookTorranceMaterialNames,
            COOK_TORRANCE_MATERIAL_NAME);

        try {
            textureMap = ImagePersistence.importRGB(new File("../../../etc/textures/miniearth.png"));

            IndexedColorImageUncompressed bump = ImagePersistence.importIndexedColor(
                new File("../../../etc/bumpmaps/earth.bw"));
            NormalMap bumpNormalMap = new NormalMap();
            bumpNormalMap.importBumpMap(bump, DEFAULT_BUMP_SCALE);
            // Keep GLSL and CPU raytracer aligned: both consume the same
            // precomputed normal field extracted from the bump map.
            bumpMapHeightRgb = bumpNormalMap.exportToRgbImage();
        }
        catch (Exception e) {
            throw new IllegalStateException("Failed loading textures for ShadersExample", e);
        }

        animationEnabled = false;
        lightAnimationEnabled = false;
        renderingMode = ShaderOperationMode.OPENGL_4_1;
        showHud = true;
        sphereRotationAngleRadians = 0.0;
        sphereMeridians = 100;
        sphereParallels = 50;
        updateSoftwareViewportAndCamera(
            (int)camera.getViewportXSize(),
            (int)camera.getViewportYSize());
    }

    public Camera getCamera()
    {
        return camera;
    }

    public CameraController getCameraController()
    {
        return cameraController;
    }

    public RendererConfiguration getQuality()
    {
        return quality;
    }

    public RendererConfigurationController getQualityController()
    {
        return qualityController;
    }

    public Sphere getSphere()
    {
        return sphere;
    }

    public Light getLight()
    {
        return light;
    }

    public SimpleMaterial getMaterial()
    {
        return material;
    }

    public SimpleMaterial getActiveMaterialForCurrentShading()
    {
        if ( quality.getShadingType() == RendererConfiguration.SHADING_TYPE_COOK_TERRANCE ) {
            return cookTorranceCopperMaterial;
        }
        return material;
    }

    public void cycleCookTorranceMaterial()
    {
        if ( cookTorranceMaterialNames == null || cookTorranceMaterialNames.isEmpty() ) {
            return;
        }
        cookTorranceMaterialIndex =
            (cookTorranceMaterialIndex + 1) % cookTorranceMaterialNames.size();
        String nextMaterialName = cookTorranceMaterialNames.get(cookTorranceMaterialIndex);
        cookTorranceCopperMaterial = new MicroFacetedMaterial(
            MICROFACET_CSV_FILE,
            nextMaterialName);
    }

    public String getCookTorranceMaterialLabel()
    {
        if ( cookTorranceCopperMaterial == null ||
             cookTorranceCopperMaterial.getName() == null ||
             cookTorranceCopperMaterial.getName().isBlank() ) {
            return COOK_TORRANCE_MATERIAL_NAME;
        }
        return cookTorranceCopperMaterial.getName();
    }

    public RGBImageUncompressed getTextureMap()
    {
        return textureMap;
    }

    public RGBImageUncompressed getBumpMapHeightRgb()
    {
        return bumpMapHeightRgb;
    }

    public RGBImageUncompressed getSoftwareFrameImage()
    {
        return softwareFrameImage;
    }

    public ShaderOperationMode getRenderingMode()
    {
        return renderingMode;
    }

    public void setRenderingMode(ShaderOperationMode renderingMode)
    {
        this.renderingMode = renderingMode;
    }

    public void rotateRenderingMode()
    {
        renderingMode = renderingMode.next();
    }

    public boolean isShowHud()
    {
        return showHud;
    }

    public void setShowHud(boolean showHud)
    {
        this.showHud = showHud;
    }

    public void toggleShowHud()
    {
        showHud = !showHud;
    }

    public void updateSoftwareViewportAndCamera(int viewportWidth, int viewportHeight)
    {
        int width = Math.max(1, viewportWidth);
        int height = Math.max(1, viewportHeight);
        camera.updateViewportResize(width, height);
        if ( softwareFrameImage != null &&
             softwareFrameImage.getXSize() == width &&
             softwareFrameImage.getYSize() == height ) {
            return;
        }
        softwareFrameImage = new RGBImageUncompressed();
        if ( !softwareFrameImage.init(width, height) ) {
            throw new IllegalStateException(
                "Could not allocate software frame image " + width + "x" + height);
        }
    }

    public boolean isAnimationEnabled()
    {
        return animationEnabled;
    }

    public void setAnimationEnabled(boolean animationEnabled)
    {
        this.animationEnabled = animationEnabled;
    }

    public void toggleAnimationEnabled()
    {
        animationEnabled = !animationEnabled;
    }

    public boolean isLightAnimationEnabled()
    {
        return lightAnimationEnabled;
    }

    public void setLightAnimationEnabled(boolean lightAnimationEnabled)
    {
        this.lightAnimationEnabled = lightAnimationEnabled;
    }

    public void toggleLightAnimationEnabled()
    {
        lightAnimationEnabled = !lightAnimationEnabled;
    }

    public double getSphereRotationAngleRadians()
    {
        return sphereRotationAngleRadians;
    }

    public void setSphereRotationAngleRadians(double sphereRotationAngleRadians)
    {
        this.sphereRotationAngleRadians = normalizeAngleRadians(sphereRotationAngleRadians);
    }

    public void advanceSphereRotationRadians(double deltaRadians)
    {
        setSphereRotationAngleRadians(sphereRotationAngleRadians + deltaRadians);
    }

    public int getSphereMeridians()
    {
        return sphereMeridians;
    }

    public int getSphereParallels()
    {
        return sphereParallels;
    }

    public void setSphereMeridians(int sphereMeridians)
    {
        this.sphereMeridians = Math.max(MIN_SPHERE_MERIDIANS, sphereMeridians);
    }

    public void setSphereParallels(int sphereParallels)
    {
        this.sphereParallels = Math.max(MIN_SPHERE_PARALLELS, sphereParallels);
    }

    public void changeSphereMeridians(int delta)
    {
        setSphereMeridians(sphereMeridians + delta);
    }

    public void changeSphereParallels(int delta)
    {
        setSphereParallels(sphereParallels + delta);
    }

    private static double normalizeAngleRadians(double angle)
    {
        double twoPi = 2.0 * Math.PI;
        double normalized = angle % twoPi;
        if ( normalized < 0.0 ) {
            normalized += twoPi;
        }
        return normalized;
    }

    private static int indexOfMaterialName(
        List<String> materialNames,
        String targetName)
    {
        int i;
        for ( i = 0; i < materialNames.size(); i++ ) {
            if ( materialNames.get(i).equalsIgnoreCase(targetName) ) {
                return i;
            }
        }
        return 0;
    }

    private static List<String> loadMicroFacetMaterialNames(String csvPath)
    {
        ArrayList<String> materialNames = new ArrayList<String>();
        File csvFile = new File(csvPath);
        try {
            List<String> lines = Files.readAllLines(
                csvFile.toPath(),
                StandardCharsets.UTF_8);
            if ( lines.isEmpty() ) {
                materialNames.add(COOK_TORRANCE_MATERIAL_NAME);
                return materialNames;
            }
            int i;
            for ( i = 1; i < lines.size(); i++ ) {
                String line = lines.get(i).trim();
                if ( line.isEmpty() ) {
                    continue;
                }
                String[] fields = line.split(",", -1);
                if ( fields.length == 0 ) {
                    continue;
                }
                String materialName = fields[0].trim();
                if ( materialName.isEmpty() ) {
                    continue;
                }
                materialNames.add(materialName);
            }
        }
        catch ( IOException e ) {
            materialNames.clear();
        }
        if ( materialNames.isEmpty() ) {
            materialNames.add(COOK_TORRANCE_MATERIAL_NAME);
        }
        return materialNames;
    }
}

File: cpp/testsuite/OpenGL4Examples/ShadersExample/src/model/ShadersModel.cpp
#include "ShadersModel.h"

#include <algorithm>
#include <cmath>
#include <cstdio>
#include "java/util/ArrayList.txx"

#include "vsdk/toolkit/common/color/ColorRgb.h"
#include "vsdk/toolkit/common/linealAlgebra/Matrix4x4d.h"
#include "vsdk/toolkit/common/linealAlgebra/Vector3Dd.h"
#include "vsdk/toolkit/environment/light/LightType.h"
#include "vsdk/toolkit/io/image/ImagePersistence.h"
#include "java/io/File.h"
#include "vsdk/toolkit/media/IndexedColorImageUncompressed.h"

ShadersModel ShadersModel::createDefault() { return ShadersModel(); }

ShadersModel::ShadersModel()
    : camera(new Camera()),
      cameraController(0),
      qualityController(0),
      sphere(new Sphere(1.0)),
      light(new Light(LightType::POINT, Vector3Dd(1, -3, 1), ColorRgb(1, 1, 1))),
      cookTorranceMaterial(0),
      cookTorranceMaterialIndex(0),
      textureMap(0), bumpMapHeightRgb(0), bumpNormalMap(0), softwareFrameImage(0),
      renderingMode(ShaderOperationMode::OPENGL_4_1), showHud(true),
      animationEnabled(false), lightAnimationEnabled(false), sphereRotationAngleRadians(0.0),
      sphereMeridians(100), sphereParallels(50)
{
    camera->setPosition(Vector3Dd(0, -4, 0));
    camera->setRotation(Matrix4x4d().eulerAnglesRotation(M_PI / 2.0, 0.0, 0.0));
    camera->setFov(30.0);

    cameraController = new CameraControllerAquynza(camera);
    quality.setTexture(true);
    quality.setBumpMap(true);
    quality.setShadingType(RendererConfiguration::SHADING_TYPE_PHONG);
    qualityController = new RendererConfigurationController(&quality);

    light->setId(0);
    material = material.withAmbient(ColorRgb(0.1, 0.1, 0.1));
    material = material.withDiffuse(ColorRgb(1, 1, 1));
    material = material.withSpecular(ColorRgb(1, 1, 1));
    material = material.withPhongExponent(40);
    cookTorranceMaterial = new MicroFacetedMaterial(
        "../../../../etc/materials/microFacetMAterials.csv",
        "Copper");

    FILE* csv = fopen("../../../../etc/materials/microFacetMAterials.csv", "r");
    if ( csv ) {
        char lineStr[4096];
        if ( fgets(lineStr, sizeof(lineStr), csv) ) {
            while ( fgets(lineStr, sizeof(lineStr), csv) ) {
                java::String line(lineStr);
                if ( line.empty() ) continue;
                java::ArrayList<java::String> cols;
                int prev = 0;
                for (int i = 0; i <= (int)line.size(); i++) {
                    if (i == (int)line.size() || line[i] == ',') {
                        cols.add(line.substr(prev, i - prev));
                        prev = i + 1;
                    }
                }
                if ( cols.size() < 16 ) continue;
                const java::String& name = cols[0];
                if ( name.empty() ) continue;
                cookTorranceMaterialNames.add(name);
                if ( name == "Copper" ) {
                    cookTorranceMaterialIndex = (int)cookTorranceMaterialNames.size() - 1;
                }
            }
        }
        fclose(csv);
    }

    try {
        java::File textureFile("../../../../etc/textures/miniearth.png");
        textureMap = ImagePersistence::importRGB(textureFile);
        java::File bumpFile("../../../../etc/bumpmaps/earth.bw");
        IndexedColorImageUncompressed* bump = ImagePersistence::importIndexedColor(bumpFile);
        bumpNormalMap = new NormalMap();
        bumpNormalMap->importBumpMap(bump, Vector3Dd(1.0, 1.0, 1.0));
        bumpMapHeightRgb = bumpNormalMap->exportToRgbImage();
        delete bump;
    }
    catch (...) {
    }

    updateSoftwareViewportAndCamera(1100, 900);
}

ShadersModel::~ShadersModel()
{
    delete qualityController;
    delete cameraController;
    delete light;
    delete sphere;
    delete textureMap;
    delete bumpMapHeightRgb;
    delete bumpNormalMap;
    delete cookTorranceMaterial;
    delete softwareFrameImage;
    delete camera;
}

void ShadersModel::rotateRenderingMode() { renderingMode = nextShaderOperationMode(renderingMode); }
void ShadersModel::toggleShowHud() { showHud = !showHud; }
void ShadersModel::toggleAnimationEnabled() { animationEnabled = !animationEnabled; }
void ShadersModel::toggleLightAnimationEnabled() { lightAnimationEnabled = !lightAnimationEnabled; }
void ShadersModel::changeSphereMeridians(int delta) { sphereMeridians = std::max(12, sphereMeridians + delta); }
void ShadersModel::changeSphereParallels(int delta) { sphereParallels = std::max(8, sphereParallels + delta); }

static double normalizeAngle(double a) {
    const double twoPi = 2.0 * M_PI;
    double n = std::fmod(a, twoPi);
    if (n < 0.0) n += twoPi;
    return n;
}

void ShadersModel::setSphereRotationAngleRadians(double angle) { sphereRotationAngleRadians = normalizeAngle(angle); }
void ShadersModel::advanceSphereRotationRadians(double delta) { setSphereRotationAngleRadians(sphereRotationAngleRadians + delta); }
const SimpleMaterial& ShadersModel::getActiveMaterialForCurrentShading() const
{
    if ( quality.getShadingType() == RendererConfiguration::SHADING_TYPE_COOK_TERRANCE &&
         cookTorranceMaterial != 0 ) {
        return *cookTorranceMaterial;
    }
    return material;
}

SimpleMaterial* ShadersModel::createActiveMaterialCopy() const
{
    if ( quality.getShadingType() == RendererConfiguration::SHADING_TYPE_COOK_TERRANCE &&
         cookTorranceMaterial != 0 ) {
        return new MicroFacetedMaterial(*cookTorranceMaterial);
    }
    return new SimpleMaterial(material);
}

void ShadersModel::updateSoftwareViewportAndCamera(int width, int height)
{
    const int w = std::max(1, width);
    const int h = std::max(1, height);
    camera->updateViewportResize(w, h);
    if (softwareFrameImage && softwareFrameImage->getXSize() == w && softwareFrameImage->getYSize() == h) return;
    delete softwareFrameImage;
    softwareFrameImage = new RGBImageUncompressed();
    softwareFrameImage->init(w, h);
}

java::String ShadersModel::getCookTorranceMaterialLabel() const
{
    if ( cookTorranceMaterial == 0 || cookTorranceMaterial->getName().empty() ) return "Copper";
    return cookTorranceMaterial->getName();
}

void ShadersModel::cycleCookTorranceMaterial()
{
    if ( cookTorranceMaterialNames.size() == 0 ) return;
    cookTorranceMaterialIndex = (cookTorranceMaterialIndex + 1) % (int)cookTorranceMaterialNames.size();
    java::String name = cookTorranceMaterialNames.get((long int)cookTorranceMaterialIndex);
    delete cookTorranceMaterial;
    cookTorranceMaterial = new MicroFacetedMaterial(
        "../../../../etc/materials/microFacetMAterials.csv",
        name);
}
